### ISSUE TITLE
Clarify the order of memory.copy / table.copy

### ISSUE THREAD

**Issue by binji** [WebAssembly/bulk-memory-operations#151]
It's not immediately obvious whether the immediates are in "dest source" or "source dest" order. For the bulk-memory proposal it doesn't matter, since there's only one memory and one table anyway. But the reference-types proposal adds mutliple tables, and the multi-memory proposal adds multiple memories, so it would be good to clarify this here.

(reported by @rreverser)

**Comment by gahaas** [comment]
This was already discussed in https://github.com/WebAssembly/bulk-memory-operations/issues/117. The order is clarified in the overview, see https://github.com/WebAssembly/bulk-memory-operations/blob/master/proposals/bulk-memory-operations/Overview.md#instruction-encoding

**Comment by RReverser** [comment]
@gahaas Good to know, thanks. It would be still useful, however, to clarify it in the spec text itself. Information from Overview won't be shown as part of the spec when merged, and the note in e.g. binary encoding section currently says only:

> **Note**
In future versions of WebAssembly, the additional zero bytes occurring in the encoding of the memory.size, memory.grow, memory.copy, and memory.fill instructions may be used to index additional memories.

**Comment by dubiousjim** [comment]
What about `memory.init`? The part of the spec that @gahaas links to implies that the order of immediates for `memory.init` is *first*, the  index of the `data` from which bytes are sourced (this immediate argument is mandatory), and *second*, the index of the destination `memory` (which may be > 0 if multi-memory proposal is enabled, or may be omitted in which case it will default to 0).

Is that really what's intended? Is it finalized?

One shortcoming of that order is that it breaks the pattern that `memory.copy` (and `table.copy`, and Unix `memcpy`) leads one to expect: destination first, then the source of data.

Another shortcoming is that in the [reference types proposal](https://github.com/WebAssembly/reference-types/blob/master/proposals/reference-types/Overview.md), they say the order of immediate arguments for `table.init` is the reverse:

> The table.init instruction takes an additional table index as immediate.
>     table.init $x $y : [i32 i32 i32] -> []
>     iff $x : table t
>     and $y : elem t'
>     and t' <: t

In other words, first the index of the table, and second the index of the source `elem`.

**Comment by dubiousjim** [comment]
Looks like the order of `table.init` was changed so that the `elem` index comes first, and the `table` index comes second: see [#81](https://github.com/WebAssembly/reference-types/issues/81) and [#82](https://github.com/WebAssembly/reference-types/pull/82). That happened in Feb 2020. But the Overview for that proposal hasn't been updated to reflect this.

**Comment by rossberg** [comment]
I think this is resolved, since it is explicit in the reftypes spec.

**Comment by dubiousjim** [comment]
For explicitness, the order in the text syntax will also be: first immediate is the destination, second immediate is the source?

**Comment by rossberg** [comment]
Right, the only place where the dest goes second is in the _binary_ format of table.init, which can be considered a historic inconsistency that is confined there.
